mantidproject/mantid
# Abins: support Lagrange instrument

**Description of work.**

~~This work builds on the draft PR #28118 and may end up replacing it - the mechanics for 2D outputs and extra instruments are quite intertwined.~~

This work is built on top of changes that were made to support 2D instruments. Subsequently the 2D front-end and instruments were removed to be added later and reduce the scope of this (still rather large) PR. That work is continued (on top of this branch) in #30440 

Some significant technical changes in Abins were made that generally support a wider range of instruments:

- Scattering data is calculated in a loop over angles. Instruments are required to provide information about the scattering angles they sample.
- A new data container SDataByAngle is introduced which allows multi-angle scattering data to be collected. This will be especially relevant for 2D plotting, but for the Lagrange instrument also provides a clear place for summing/averaging over scattering angles to take place in a single step. Informal testing suggests that this does not introduce any performance penalty. (Actually, a small but consistent improvement in runtime was observed.)
- Instruments may have multiple operating modes. In Abins these are called "settings", while similar mechanics would be used in Abins2D for chopper packages.

An additional piece of refactoring was primarily done to support Abins2D but seems more effort than it is worth to unpick at this point. (We'll need it back soon!)

- A base class for the Abins Algorithm is provided in `abins.abinsalgorithm`. This will make it easier to maintain Abins and Abins2D concurrently, as the main load points under Framework will focus on the elements that are _different_ and common functionality is deferred to the parent class.

Specific changes to Abins for the benefit of ILL-Lagrange users are:
- The addition of a Lagrange instrument option with resolution functions for the various monochromators
- Addition of an "EnergyUnits" dialogue box to improve the workflow of users who prefer to work in meV. Under the hood everything is still done in wavenumbers, and the ConvertUnits algorithm is called at the end to convert the GroupWorkspace. Experimental data files are assumed to be given in the same units.


**To test:**
- Check unit/doc/integration tests are passing
- Run the Abins algorithm, selecting the Lagrange instrument
- Try running with different "setting" options for both TOSCA and Lagrange
  - Lagrange monochromators should broaden the data in different ways
  - TOSCA forward/backward/all options should affect the relative peak intensities
  - The blank option should select a default for the current instrument at runtime
  -  Unsupported options should be picked up by the validator with a useful tooltip.
- Try setting meV units; this should trigger the ConvertUnits algorithm and scale the results accordingly

Release notes have been added

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
